AudioDeviceExample/AudioDevices/ExampleCoreAudioDevice.m

```diff
@@ -0,0 +1,487 @@
+
+    UInt32 enableOutput = 1;
+    status = AudioUnitSetProperty(_audioUnit, kAudioOutputUnitProperty_EnableIO,
+                                  kAudioUnitScope_Output, kOutputBus,
+                                  &enableOutput, sizeof(enableOutput));
+    if (status != 0) {
+        NSLog(@"Could not enable output bus!");
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    status = AudioUnitSetProperty(_audioUnit, kAudioUnitProperty_StreamFormat,
+                                  kAudioUnitScope_Input, kOutputBus,
+                                  &streamDescription, sizeof(streamDescription));
+    if (status != 0) {
+        NSLog(@"Could not enable output bus!");
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    // Disable input, we don't want it.
+    UInt32 enableInput = 0;
+    status = AudioUnitSetProperty(_audioUnit, kAudioOutputUnitProperty_EnableIO,
+                                  kAudioUnitScope_Input, kInputBus, &enableInput,
+                                  sizeof(enableInput));
+
+    if (status != 0) {
+        NSLog(@"Could not disable input bus!");
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    // Setup the rendering callback.
+    AURenderCallbackStruct renderCallback;
+    renderCallback.inputProc = ExampleCoreAudioDevicePlayoutCallback;
+    renderCallback.inputProcRefCon = (void *)(context);
+    status = AudioUnitSetProperty(_audioUnit, kAudioUnitProperty_SetRenderCallback,
+                                  kAudioUnitScope_Output, kOutputBus, &renderCallback,
+                                  sizeof(renderCallback));
+    if (status != 0) {
+        NSLog(@"Could not set rendering callback!");
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    // Finally, initialize and start the RemoteIO audio unit.
+    status = AudioUnitInitialize(_audioUnit);
+    if (status != 0) {
+        NSLog(@"Could not initialize the audio unit!");
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+        return NO;
+    }
+
+    return YES;
+}
+
+- (BOOL)startAudioUnit {
+    OSStatus status = AudioOutputUnitStart(_audioUnit);
+    if (status != 0) {
+        NSLog(@"Could not start the audio unit!");
+        return NO;
+    }
+    return YES;
+}
+
+- (BOOL)stopAudioUnit {
+    OSStatus status = AudioOutputUnitStop(_audioUnit);
+    if (status != 0) {
+        NSLog(@"Could not stop the audio unit!");
+        return NO;
+    }
+    return YES;
+}
+
+- (void)teardownAudioUnit {
+    if (_audioUnit) {
+        AudioUnitUninitialize(_audioUnit);
+        AudioComponentInstanceDispose(_audioUnit);
+        _audioUnit = NULL;
+    }
+}
+
+- (void)registerAVAudioSessionObservers {
```

Comment:
This method is not limited to observing AVAudioSessionan, it also add an observer to `UIApplicationDidBecomeActiveNotification`.  Should we change the name to generic

Reply:
I think we can keep it as is. The only "leakage" outside of AVAudioSession is the iOS 9 workaround.